File: Lintcode/46_majority-number/majority-number.cpp
/*
@Copyright:LintCode
@Author:   mts
@Problem:  http://www.lintcode.com/problem/majority-number
@Language: C++
@Datetime: 15-09-30 21:29
*/

class Solution {
public:
    /**
     * @param nums: A list of integers
     * @return: The majority number
     */
    int majorityNumber(vector<int> nums) {
        // write your code here
        int majIndex = 0;
        int c = 1;
        
        for(int i=0; i<nums.size(); i++){
            if(nums[majIndex] == nums[i]){
                c++;
            }
            else{
                c--;
            }
            if(c == 0){
                // Reset the counter
                majIndex = i;
                c = 1;
            }
        }
        
        return nums[majIndex];
    }
};

File: Lintcode/46_majority-number/majority-number.java
/*
@Copyright:LintCode
@Author:   mts
@Problem:  http://www.lintcode.com/problem/majority-number
@Language: Java
@Datetime: 16-11-26 03:05
*/

public class Solution {
    /**
     * @param nums: a list of integers
     * @return: find a  majority number
     */
    public int majorityNumber(ArrayList<Integer> nums) {
        int n = nums.size();
        if(n == 0)
            return 0;
        int cnt = 1;
        int x = nums.get(0);
        int i = 1;
        while(i < n){
            if(nums.get(i) == x)
                cnt++;
            else{
                cnt--;
                if(cnt == 0){
                    x = nums.get(i);
                    cnt = 1;
                }
            }
            i++;
        }
        return x;
   }
}